homeassistant/components/sma/config_flow.py

```diff
@@ -100,3 +123,49 @@ async def async_step_user(
             ),
             errors=errors,
         )
+
+    async def async_step_dhcp(
+        self, discovery_info: DhcpServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle DHCP discovery."""
+        _LOGGER.debug("DHCP discovery: %s", discovery_info)
+        self._discovery_data[CONF_HOST] = discovery_info.ip
+        self._discovery_data[CONF_MAC] = format_mac(discovery_info.macaddress)
+        self._discovery_data[CONF_NAME] = discovery_info.hostname.upper()
+
+        return await self.async_step_discovery_confirm()
```

Comment:
The hostname contains the unique id, let's already have a thing here that sets the unique id of the flow, and aborts (and updates the existing data) if we already have that unique id registered

Reply:
Don't forget to set `await self.async_set_unique_id(str(device_info["serial"]))` with `raise_on_progress=False` in the `async_step_user`, as that will make sure if HA discovered the device, and the user does not pick the discovery, but just does it manually, it is able to finish (and abort the discovered one)